### PR TITLE
Refactor ServerQueryRequest to pre-compute and store segment independent information about the request

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Compile all expressions in aggregation and group-by
Extract columns in filter, aggregation, group-by and filter
Do not cache InstanceRequest
Reduce the work in segment pruner from per segment basis to only once

Next step: use ServerQueryRequest for the query execution instead of BrokerRequest

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/request/RequestUtils.java [hunk 1]
```diff
@@ -149,18 +151,58 @@ public static FilterQueryTree buildFilterQuery(Integer id, Map<Integer, FilterQu
...
+    Set<String> filterColumns = new HashSet<>();
+    if (root.getChildren() == null) {
+      filterColumns.add(root.getColumn());
+    } else {
+      Stack<FilterQueryTree> stack = new Stack<>();
+      stack.add(root);
+      while (!stack.empty()) {
+        FilterQueryTree node = stack.pop();
+        for (FilterQueryTree child : node.getChildren()) {
+          if (child.getChildren() == null) {
```
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/request/RequestUtils.java]: check for both null and empty?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java [hunk 2]
```diff
@@ -15,111 +15,38 @@
...
-          }
-        }
-      }
-      return false;
-    }
-    // unsupported query type.
-    return true;
+  @Override
+  public boolean prune(@Nonnull IndexSegment segment, @Nonnull ServerQueryRequest queryRequest) {
+    Set<String> ColumnsInSchema = segment.getSegmentMetadata().getSchema().getColumnNames();
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java]: ColumnsInSchema --> columnsInSchema

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java [hunk 3]
```diff
@@ -15,111 +15,38 @@
...
+    return !(containsAll(ColumnsInSchema, queryRequest.getFilterColumns()) && containsAll(ColumnsInSchema,
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java]: should we not just construct columnsInQuery and do one check against that?

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/request/RequestUtils.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/request/RequestUtils.java]: **children** is non-null only for AND and OR, and will never be empty

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java]: Fixed

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/DataSchemaSegmentPruner.java]: Done

**Comment by Jackie-Jiang** [comment]
@kishoreg For the **Nonnull** and **Nullable** annotations, it's for the clarity of the public APIs. With the annotations, when we call the methods, we know whether it expects a null value, and thus prevent causing NPE.
Also, IntelliJ can give out warnings when it detects potential problems: https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html

**Pull request merged**